src/garage/torch/algos/vpg.py

```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
             discount=self.discount)
         return np.mean(undiscounted_returns)
 
-    def _train(self, obs, actions, rewards, advantages):
-        r"""Train the algorithm with minibatch.
+    def _train(self, obs_flat, actions_flat, rewards_flat, returns_flat,
+               advs_flat):
+        r"""Train the policy and value function with minibatch.
+
+        Args:
+            obs_flat (torch.Tensor): Observation from the environment
+                with shape :math:`(N, O*)`.
+            actions_flat (torch.Tensor): Actions fed to the environment
+                with shape :math:`(N, A*)`.
+            rewards_flat (torch.Tensor): Acquired rewards
+                with shape :math:`(N, )`.
+            returns_flat (torch.Tensor): Acquired returns
+                with shape :math:`(N, )`.
+            advs_flat (torch.Tensor): Advantage value at each step
+                with shape :math:`(N, )`.
+
+        """
+        dataset = BatchDataset(
+            (obs_flat, actions_flat, rewards_flat, returns_flat, advs_flat),
+            self._minibatch_size)
+
+        for epoch in range(self._optimization_epochs):
+            for obs, actions, rewards, returns, advs in dataset.iterate():
+                policy_loss = self._train_policy(obs, actions, rewards, advs)
+                vf_loss = self._train_value_function(obs, returns)
```

Comment:
it's not clear that it's algorithmically valid to modify the value function during policy training. if you are feeding act/obs through the value function to calculate predicted returns, these predicted returns will start using information from the current batch to calculate predicted returns, which is not valid. the baseline should only have information about the batches prior to this one.

where did you get the idea of doing this (vs just optimizing it after the policy)?

Reply:
I don't know this is related to above comments what you post, but I thought every value (advantages, baselines, returns, ...) are computed at the beginning of the training process in the `process_samples` function. So even I update the value_function with the policy, I don't need to re-compute the values(advantages and baselines) and it would be okay.

Reply:
ok second read, this seems probably okay to me as long as `tu.discount_cumsum` is making differentiable operations.

are you sure that if this algorithm were moved to the GPU, that all of these tensors would go with it?

Reply:
As I understand, I have to convert tensors using CUDA commands like 
```
if torch.cuda.is_availabe():
    tensors = tensors.cuda()
```
Are you talking about this codes?

Reply:
i am asking whether all of these operations are performed as part of the torch computation graph.